sys/picolibc_syscalls_default/syscalls.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @defgroup    sys_picolibc PicoLibc system call
+ * @ingroup     sys
+ * @brief       PicoLibc system call
+ * @{
+ *
+ * @file
+ * @brief       PicoLibc system call implementations
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include "log.h"
+#include "periph/pm.h"
+#include "stdio_base.h"
+
+uint8_t sbrk[1];
+
+/**
+ * @brief Exit a program without cleaning up files
+ *
+ * If your system doesn't provide this, it is best to avoid linking with subroutines that
+ * require it (exit, system).
+ *
+ * @param n     the exit code, 0 for all OK, >0 for not OK
+ */
+void _exit(int n)
+{
+    LOG_INFO("#! exit %i: powering off\n", n);
+    pm_off();
+    while(1);
+}
+
+/**
+ * @brief Send a signal to a thread
+ *
+ * @param[in] pid the pid to send to
+ * @param[in] sig the signal to send
+ *
+ * @return    always returns -1 to signal error
+ */
+__attribute__ ((weak))
+int kill(pid_t pid, int sig)
+{
+    (void) pid;
+    (void) sig;
+    errno = ESRCH;                         /* not implemented yet */
+    return -1;
+}
+
+/* File descriptor pointers */
+int picolibc_get(FILE *file)
+{
+    (void)file;
+    char c = 0;
+    stdio_read(&c, 1);
+    return c;
+}
+
+int picolibc_put(char c, FILE *file)
+{
+    (void)file;
+    stdio_write(&c, 1);
+    return 1;
+}
+FILE picolibc_stdio =
+    FDEV_SETUP_STREAM(picolibc_put, picolibc_get, NULL, _FDEV_SETUP_RW);
+
+FILE *const __iob[] = {
+    &picolibc_stdio,
+    &picolibc_stdio,
+    &picolibc_stdio,
+};
+
+/*
+ * All output is directed to stdio_uart, independent of the given file descriptor.
+ * The write call will further block until the byte is actually written to the UART.
+ */
+_READ_WRITE_RETURN_TYPE write(int fd, const void *data, size_t count)
+{
+    (void) fd;
+    return stdio_write(data, count);
+}
```

Comment:
Can't we just call `vfs_write()` here?

Reply:
> `picolibc_syscalls_default/syscalls.c` and `newlib_syscalls_default/syscalls.c` can probably share some code, e.g. no need to implement `_sbrk_r` twice.

A couple of thoughts here:

 * picolibc provides a version of sbrk that uses all remaining RAM, if you want to use that.
 * picolibc doesn't need _sbrk_r, it just wants regular sbrk. You could save a bit of space by refactoring to eliminate the extra function.

Reply:
Will picolibc have problems if the chunks returned by `sbrk` are not continuous?
E.g. some MCUs have non-continuous blocks of RAM, unusually intended for a special function, but if that's not used we just add that to the heap.
We noticed that newlib does something clever internally that breaks if there are gaps in the addresses `_sbrk_r` returns.
With newlib-nano it would just work, so I guess picolib wouldn't do anything special here either.

Reply:
Picolibc is just using the nano-mallocr from newlib, which looks like it will be fine with non-contiguous blocks returned from sbrk. The more complicated regular malloc in newlib might well have troubles. I think the only issue with non-continuous blocks is that mallinfo will return bogus information.